File: sem4/aisd/l1/z1/Queue.java
public class Queue {
    private int[] queue;
    private int front;
    private int rear;
    private int capacity;
    private int count;

    public Queue(int size) {
        queue = new int[size];
        capacity = size;
        front = 0;
        rear = -1;
        count = 0;
    }

    public void enqueue(int item) {
        if (isFull()) {
            System.out.println("Queue is full");
            System.exit(1);
        }

        rear = (rear + 1) % capacity;
        queue[rear] = item;
        count++;
    }

    public int dequeue() {
        if (isEmpty()) {
            System.out.println("Queue is empty");
            System.exit(1);
        }

        int item = queue[front];
        front = (front + 1) % capacity;
        count--;
        return item;
    }
    
    public int size() {
        return count;
    }

    public boolean isEmpty() {
        return (size() == 0);
    }

    public boolean isFull() {
        return (size() == capacity);
    }
}

File: mk1/AISD/Lista1/structures/queue.go
package structures

import "errors"

type Queue[T any] struct {
	memory []T
}

func (q *Queue[T]) Enqueue(item T) {
	q.memory = append(q.memory, item)
}

func (q *Queue[T]) Dequeue() (T, error) {
	if len(q.memory) == 0 {
		var zero T
		return zero, errors.New("empty queue")
	}
	ret := q.memory[0]
	q.memory = q.memory[1:]
	return ret, nil
}
